File: C/mergesort/qwen-qwq-32b.c
#include <stdio.h>
#include <stdlib.h>

#define INSERTION_SORT_THRESHOLD 15

void insertionSort(int arr[], int left, int right) {
    for (int i = left + 1; i <= right; i++) {
        int key = arr[i];
        int j = i - 1;
        while (j >= left && arr[j] > key) {
            arr[j+1] = arr[j];
            j--;
        }
        arr[j+1] = key;
    }
}

void merge(int arr[], int left, int mid, int right, int temp[]) {
    // Copy elements to temp
    for (int i = left; i <= right; i++) {
        temp[i] = arr[i];
    }

    int i = left;
    int j = mid + 1;
    int k = left;

    while (i <= mid && j <= right) {
        if (temp[i] <= temp[j]) {
            arr[k] = temp[i];
            i++;
        } else {
            arr[k] = temp[j];
            j++;
        }
        k++;
    }

    while (i <= mid) {
        arr[k] = temp[i];
        i++;
        k++;
    }

    while (j <= right) {
        arr[k] = temp[j];
        j++;
        k++;
    }
}

void mergeSort(int arr[], int left, int right, int temp[]) {
    if (right - left + 1 < INSERTION_SORT_THRESHOLD) {
        insertionSort(arr, left, right);
        return;
    }
    if (left < right) {
        int mid = left + (right - left)/2;
        mergeSort(arr, left, mid, temp);
        mergeSort(arr, mid+1, right, temp);
        merge(arr, left, mid, right, temp);
    }
}

void mergeSortWrapper(int arr[], int n) {
    if (n <= 1) return;
    int *temp = (int*)malloc(n * sizeof(int));
    if (temp == NULL) {
        // handle error
        return;
    }
    mergeSort(arr, 0, n-1, temp);
    free(temp);
}

int main() {
    int arr[1000];
    int n = 0;

    FILE *file = fopen("../../inputs.txt", "r");
    if (!file) {
        return 1;
    }

    while (fscanf(file, "%d", &arr[n]) == 1 && n < 1000) {
        n++;
    }
    fclose(file);

    mergeSortWrapper(arr, n);

    return 0;
}

File: Python/mergesort/qwen-qwq-32b.py
def insertion_sort(arr, left, right):
    for i in range(left + 1, right + 1):
        key = arr[i]
        j = i - 1
        while j >= left and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key

def merge_sort(arr):
    if len(arr) < 2:
        return
    temp = [0] * len(arr)
    merge_sort_helper(arr, temp, 0, len(arr) - 1)


def merge_sort_helper(arr, temp, left, right):
    if right - left + 1 < 15:
        insertion_sort(arr, left, right)
        return
    if left < right:
        mid = (left + right) // 2
        merge_sort_helper(arr, temp, left, mid)
        merge_sort_helper(arr, temp, mid + 1, right)
        merge(arr, temp, left, mid, right)

def merge(arr, temp, left, mid, right):
    # Check if already sorted
    if arr[mid] <= arr[mid + 1]:
        return
    # Copy to temp array
    for i in range(left, right + 1):
        temp[i] = arr[i]
    a = left
    b = mid + 1
    curr = left
    while a <= mid and b <= right:
        if temp[a] <= temp[b]:
            arr[curr] = temp[a]
            a += 1
        else:
            arr[curr] = temp[b]
            b += 1
        curr += 1
    # Copy remaining elements from left
    while a <= mid:
        arr[curr] = temp[a]
        a += 1
        curr += 1
    
import sys

def main():
    input_path = "inputs.txt" if len(sys.argv) < 2 else sys.argv[1]

    try:
        with open(input_path, 'r') as f:
            arr = [int(line.strip()) for line in f if line.strip()]
    except FileNotFoundError:
        return

    sorted_arr = merge_sort(arr)

if __name__ == "__main__":
    main()